File: code/revolver.cpp
#include "revolver.h"
#include <cmath>

const int twist_limit = 5;

Revolver::Revolver():current(0),holders(5,HOLDER_OPEN){}

int Revolver::clockwiseLimit()const{return twist_limit-current;}
int Revolver::anticlockwiseLimit()const{return -twist_limit+current;}

Holder Revolver::currentBuoy()const{
    return holders[current % 5];
}
bool Revolver::isFull()const{
    for(Holder b:holders){
        if(b==HOLDER_OPEN)return false;
    }
    return true;
}
int Revolver::nextFree()const{
    int shortestPath = 100;
    for(int i=anticlockwiseLimit();i<=clockwiseLimit();++i){
        if(std::abs(i)<std::abs(shortestPath)&&
            (holders[(current+i)%5]==HOLDER_OPEN || holders[(current+i)%5]==HOLDER_WAITING))
            shortestPath = i;
    }
    return shortestPath;
}
int Revolver::nextRed()const{
    int shortestPath = 100;
    for(int i=anticlockwiseLimit();i<=clockwiseLimit();++i){
        if(std::abs(i)<std::abs(shortestPath)&&
            (holders[(current+i)%5]==HOLDER_RED))
            shortestPath = i;
    }
    return shortestPath;
}
int Revolver::nextGreen()const{
    int shortestPath = 100;
    for(int i=anticlockwiseLimit();i<=clockwiseLimit();++i){
        if(std::abs(i)<std::abs(shortestPath)&&
            (holders[(current+i)%5]==HOLDER_GREEN))
            shortestPath = i;
    }
    return shortestPath;
}
int Revolver::nextBuoy()const{
    int shortestPath = 100;
    for(int i=anticlockwiseLimit();i<=clockwiseLimit();++i){
        if(std::abs(i)<std::abs(shortestPath)&&
            (holders[(current+i)%5]==HOLDER_RED || holders[(current+i)%5]==HOLDER_GREEN))
            shortestPath = i;
    }
    return shortestPath;
}

void Revolver::rotate(int count) {
    current += count;
}
int Revolver::totalRed()const{
    int current_totalRed = 0;
    for(int i=0; i<5; i++) {
        if (holders[i] == HOLDER_RED) {
            current_totalRed += 1;
        }
    }
    return current_totalRed;
}
int Revolver::totalGreen()const{
    int current_totalGreen = 0;
    for (int i=0;i<5;i++) {
        if (holders[i] == HOLDER_GREEN) {
            current_totalGreen += 1;
        }
    }
    return current_totalGreen;
}
bool Revolver::isEmpty()const{
    for (int i=0;i<5;i++) {
        if (holders[i] == HOLDER_GREEN || holders[i] == HOLDER_RED) {
            return false;
        }
    }
    return true;
}
void Revolver::addCup(bool red) {
    Holder cup;
    if (red == false) {
        cup = HOLDER_GREEN;
    }
    else {
        cup = HOLDER_RED;
    }
    holders[current%5] = cup;
}
void Revolver::setHolder(int n,Holder h){
    if(n==-1)holders[current%5]=h;
    else holders[n%5]=h;
}

std::ostream& operator<<(std::ostream& s,const Revolver& r){
    for(int i=0;i<5;++i){
        if(r.current%5 == i) s<<"~~";
        switch(r.holders[i]){
            case HOLDER_OPEN: s<<"OPEN";break;
            case HOLDER_WAITING: s<<"WAITING";break;
            case HOLDER_CLOSED: s<<"CLOSED";break;
            case HOLDER_RED: s<<"RED";break;
            case HOLDER_GREEN: s<<"GREEN";break;
        }
        if(r.current%5 == i) s<<"~~";
        if(i<4) s<<" | ";
    }
    return s;
}

File: code/py/revolver.py
import motors,servos

"""
Functions for manipulating revolver
"""

# Either empty string (empty and closed), "OPEN", "RED" or "GREEN"
holders = ["","","",""]

# Index into holders
current = 0

def redCount():
    total = 0
    for h in holders:
        if h == "RED":
            total += 1

    return total

def freeCount():
    free_counter = 0 
    for holder in holders:
        if holder == "OPEN" or holder == "":
            free_counter = free_counter + 1
            
    return free_counter

def greenCount():
    greens = 0
    for holder in holders:
        if holder == "GREEN":
            greens += 1
    return greens

def count():
    return greenCount() + redCount()

def isFull():
    return (count() == 4)

def isEmpty():
    if count() == 0:
        return True
    return False

"""
True if currently active slot has a buoy in it
"""
def currentFull():
    if holders[current] == "GREEN" or holders[current] == "RED":
        return True
    return False

"""
True if holder i has a buoy in it
"""
def holderFull(i):
    if holders[i] == "" or holders[i] == "OPEN":
        return False
    return True

"""
Returns the index of a nearby empty holder, or None if all are full
"""
def findFreeSpace():
    if isFull() == True:
        return None
    if currentFull() == False:
        return current
    if not holderFull((current+1) % 4):
        return (current+1) % 4
    if not holderFull((current-1) % 4):
        return (current-1) % 4
    return (current+2) % 4

"""
Rotate to holder i
"""
def goToIndex(i):
    turnRevolver(i - current)

"""
Rotates to a nearby free space

True/False on success/failure
"""
def goToFreeSpace():
    i = findFreeSpace()
    if i is None:
        return False
    setHolder(i,"RELEASED")
    goToIndex(i)
    return True

"""
Rotates to a nearby buoy of the specified colour

True on success, False if there are no buoys of the specified colour
"""
def goToColour(colour):
    for x in range(len(holder)):
        if holders[x] == colour:
            goToIndex(x)
            return True
    return False

def goToGreen():
    return goToColour("GREEN")

def goToRed():
    return goToColour("RED")

"""
release current holder
"""
def release():
    servos.setHolder(current, "RELEASE")

"""
grab the current holder, setting its colour to red if isRed is True,
green if False, and empty+closed if None
"""
def grab(isRed):
    servos.setHolder(current, "GRABBED")
    if isRed is None:
        holders[current] = ""
    elif isRed:
        holders[current] = "RED"
    else:
        holders[current] = "GREEN"
